### PR TITLE
[ThirdParty] More NCCL tests (moved from another PR)

### PR DESCRIPTION
**Pull request by feiwang3311**

### CODE REVIEW

File: src/main/scala/lms/thirdparty/nccl/nccl.scala [hunk 1]
```diff
@@ -307,11 +307,72 @@ trait NCCLOps extends CLibs with SizeTOps with CudaOps {
...
+  /**
+   * One-to-all (scatter)
+   * A one-to-all operation from a root rank can be expressed by merging all send and receive operations in a group
+   * `root` is the root rank, `myRank` is the rank of current device, `nRanks` is the number of devices associated with
+   * the communicator `comm`.
+   */
+  def ncclScatter[T:Manifest](sendbuff: Rep[Array[Array[T]]], recvbuff: Rep[Array[T]], root: Rep[Int], myRank: Rep[Int], nRanks: Rep[Int],
+      count: Rep[Int], datatype: Rep[ncclDataTypeT], comm: Rep[ncclCommT], stream: Rep[cudaStreamT])(implicit __pos: SourceContext) = {
+    ncclCheck(ncclGroupStart())
+    __ifThenElse(equals(myRank, root), {
```
Comment (Kraks) [inline src/main/scala/lms/thirdparty/nccl/nccl.scala:332]: Doesn't `if` work here?

File: src/main/scala/lms/thirdparty/nccl/nccl.scala [hunk 2]
```diff
@@ -307,11 +307,72 @@ trait NCCLOps extends CLibs with SizeTOps with CudaOps {
...
+  }
+
+  /**
+   * All-to-one (gather)
+   * Similar to One-to-all (scatter)
+   */
+  def ncclGather[T:Manifest](sendbuff: Rep[Array[T]], recvbuff: Rep[Array[Array[T]]], root: Rep[Int], myRank: Rep[Int], nRanks: Rep[Int],
+        count: Rep[Int], datatype: Rep[ncclDataTypeT], comm: Rep[ncclCommT], stream: Rep[cudaStreamT])(implicit __pos: SourceContext) = {
+    ncclCheck(ncclGroupStart())
+    __ifThenElse(equals(myRank, root), {
```
Comment (Kraks) [inline src/main/scala/lms/thirdparty/nccl/nccl.scala:348]: same here

**Review by Kraks — APPROVED** [review]
minor comments on using `if` instead of `__ifThenElse`

File: src/main/scala/lms/thirdparty/nccl/nccl.scala (hunk 1 above)
Comment (feiwang3311) [reply src/main/scala/lms/thirdparty/nccl/nccl.scala:332]: no. We cannot use @virtualize here unfortunately.

File: src/main/scala/lms/thirdparty/nccl/nccl.scala (hunk 2 above)
Comment (feiwang3311) [reply src/main/scala/lms/thirdparty/nccl/nccl.scala:348]: same above

**Pull request merged**